### PR TITLE
Add parameter base_image and addon_image to BUILD_PLATFORMS

### LINKED ISSUE #145: Add support for windows multi-image tag builds
When doing a multiarch build (`make push-multiarch`) we might need to override additional args that might be set in the `Dockerfile.Windows` file for a multi-image build, for example

https://github.com/kubernetes-csi/node-driver-registrar/blob/bd1ad6268a014dceaa36c71c14a3fc0ad8240b58/Dockerfile.Windows#L1-L15

```Dockerfile
ARG CORE_IMAGE=servercore
ARG CORE_IMAGE_TAG=1809
ARG BUILD_IMAGE=nanoserver
ARG BUILD_IMAGE_TAG=1809
ARG REGISTRY=mcr.microsoft.com/windows

FROM ${REGISTRY}/${CORE_IMAGE}:${CORE_IMAGE_TAG} as core
FROM ${REGISTRY}/${BUILD_IMAGE}:${BUILD_IMAGE_TAG}
LABEL description="CSI Node driver registrar"

COPY ./bin/csi-node-driver-registrar.exe /csi-node-driver-registrar.exe
COPY --from=core /Windows/System32/netapi32.dll /Windows/System32/netapi32.dll

USER ContainerAdministrator
ENTRYPOINT ["/csi-node-driver-registrar.exe"]
```

We can override `BUILD_IMAGE` and `BUILD_IMAGE_TAG` before building the docker image e.g.

```
docker build \
   --build-arg BUILD_IMAGE=X \
   --build-arg BUILD_IMAGE_TAG=Y \
   -f Dockerfile.windows .
```

The existing Makefile has support for https://github.com/kubernetes-csi/csi-release-tools/blob/master/prow.sh#L80 `BUILD_PLATFORMS="linux amd64; windows amd64 .exe; ...`, in the Makefile this string is split by `;` and used in this fragment:

```
		echo "$$build_platforms" | tr ';' '\n' | while read -r os arch suffix; do \
			docker buildx build --push \
				--tag $(IMAGE_NAME):$$arch-$$os-$$tag \
				--platform=$$os/$$arch \
				--file $$(eval echo \$${dockerfile_$$os}) \
				--build-arg binary=./bin/$*$$suffix \
				--build-arg ARCH=$$arch \
				--label revision=$(REV) \
				.; \
		done; \
```

We'd like to include additional build args when calling `docker build`

---

One approach that I've seen in PD CSI is to have multiple targets per windows build https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/cfe67c8322c9094030bdf28737a50eb5b1c2a542/Makefile#L51-L78, an approach would be to split `BUILD_PLATFORMS` to `LINUX_BUILD_PLATFORMS` and `WINDOWS_BUILD_PLATFORMS` and then do:

```
LINUX_BUILD_PLATFORMS="linux ppc64le -ppc64le; linux s390x -s390x; linux arm64 -arm64 ..."
WINDOWS_BUILD_PLATFORMS="windows amd64 nanoserver:1809; windows amd64 servercore:20H2 ..."
```

		echo "$$linux_build_platforms" | tr ';' '\n' | while read -r os arch suffix; do \
			docker buildx build --push \
				--tag $(IMAGE_NAME):$$arch-$$os-$$tag \
				--platform=$$os/$$arch \
				--file $$(eval echo \$${dockerfile_$$os}) \
				--build-arg binary=./bin/$*$$suffix \
				--build-arg ARCH=$$arch \
				--label revision=$(REV) \
				.; \
		done; \
		echo "$$windows_build_platforms" | tr ';' '\n' | while read -r os arch base_image; do \
			docker buildx build --push \
				--tag $(IMAGE_NAME):$$arch-$$os-$$tag \
				--platform=$$os/$$arch \
				--file $$(eval echo \$${dockerfile_$$os}) \
				--build-arg binary=./bin/$*.exe \
                --build-arg BASE_IMAGE=$$base_image \
				--build-arg ARCH=$$arch \
				--label revision=$(REV) \
				.; \
		done; \
---

Another approach would be to add more tokens to the string (I'm not sure if at the end to make it backwards compatible) for windows e.g.

```
BUILD_PLATFORMS="windows amd64 .exe nanoserver:1909; ..."


		echo "$$build_platforms" | tr ';' '\n' | while read -r os arch suffix base_image; do \
			docker buildx build --push \
				--tag $(IMAGE_NAME):$$arch-$$os-$$tag \
				--platform=$$os/$$arch \
				--file $$(eval echo \$${dockerfile_$$os}) \
				--build-arg binary=./bin/$*$$suffix \
                --build-arg BASE_IMAGE=$$base_image \
				--build-arg ARCH=$$arch \
				--label revision=$(REV) \
				.; \
		done; \
```

---

These are just a few approaches, any feedback would be appreciated!

Also ref https://github.com/kubernetes-csi/node-driver-registrar/issues/135

cc @jingxu97

### PR DESCRIPTION
**Pull request by mauriciopoppe**
Fixes #145

Adds new parameters in BUILD_PLATFORMS referred as `base_image` and `addon_image` with the format `image:tag`

```
make push-multiarch BUILD_PLATFORMS="linux amd64; windows amd64 .exe nanoserver:2019 servercore:2019;
```

List of changes:

- made /bin/bash the default shell instead of /bin/sh, I needed to do this to do string addition and indexOf
- updated the cloudbuild docker image to a newer version that has 20.10 and this fix: https://github.com/docker/cli/issues/2722#issuecomment-693395773
- docker build with multiple build platforms with different base images
- created a docker manifest with info about all of these builds including info about the combinations of arch, os and os-version

Output of some snippets used here:

```
build_platforms=build_platforms="linux amd64; linux ppc64le -ppc64le; linux s390x -s390x; linux arm64 -arm64; windows amd64 .exe nanoserver:1809 servercore:ltsc2019; windows amd64 .exe nanoserver:1909 servercore:1909; windows amd64 .exe nanoserver:2004 servercore:2004; windows amd64 .exe nanoserver:20H2 servercore:20H2"
echo "$build_platforms" | sed -e 's/windows *[^ ]* *.exe *[^ ]* *[^ ]*//g' -e 's/; *;/;/g' -e 's/;[ ]*$//'
linux amd64; linux ppc64le -ppc64le; linux s390x -s390x; linux arm64 -arm64
```

I used these changes in node-driver-registrar to push all the windows images as well as the manifest:

image

Manifest contents of the `canary` manifest:

```json
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 738,
         "digest": "sha256:619f0eb76b5b704fe351dacc26b94368e7e0db60a1fefcd8605aeb3e7e81874e",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 950,
         "digest": "sha256:5a21cee79643052c985a0d7073bef60f36711c83d74f58e94605c3e8a1ae6abf",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.17763.1935"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 950,
         "digest": "sha256:b5f11b3fe645a67b3ae2ab3bf32403670c6c454c4ab19e72a41c052035613873",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.18363.1556"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 950,
         "digest": "sha256:fe8327a9e9d3c2c73fa117cb4b635fc2c4f3f501e217ffe4f36bb6918cb63b3e",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.19041.985"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 950,
         "digest": "sha256:fe57e7ddfcec7008290dbf998c131ac5d98c8514326555d8fb24607b17e7ca0a",
         "platform": {
            "architecture": "amd64",
            "os": "windows",
            "os.version": "10.0.19042.985"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 738,
         "digest": "sha256:cd870ccc4e58a592195e3af0cc3e690cc009a6fce7a47deaedc7583ccc335310",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 738,
         "digest": "sha256:56e373d03c5a0e00d8e09912d462984bd5169433fca887505310f4b179430386",
         "platform": {
            "architecture": "ppc64le",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 738,
         "digest": "sha256:a11755a59d704d2da641794aa31235875f1d2b65d24810831b952220fcdec86f",
         "platform": {
            "architecture": "s390x",
            "os": "linux"
         }
      }
   ]
}
```

/assign @pohly @jingxu97

### CODE REVIEW

File: build.make [hunk 1]
```diff
@@ -63,22 +63,22 @@ endif
 # Specific packages can be excluded from each of the tests below by setting the *_FILTER_CMD variables
 # to something like "| grep -v 'github.com/kubernetes-csi/project/pkg/foobar'". See usage below.
 
-# BUILD_PLATFORMS contains a set of <os> <arch> <suffix> triplets,
+# BUILD_PLATFORMS contains the tuple [os arch suffix base_image],
```
Comment (pohly) [inline build.make]: I find this confusing. It contains not just one tuple.

File: build.make [hunk 2]
```diff
@@ -143,18 +143,19 @@ $(CMDS:%=push-multiarch-%): push-multiarch-%: check-pull-base-ref build-%
 	fi; \
 	pushMultiArch () { \
 		tag=$$1; \
-		echo "$$build_platforms" | tr ';' '\n' | while read -r os arch suffix base_image; do \
+		echo "$$build_platforms" | tr ';' '\n' | while read -r os arch suffix base_image addon_image; do \
 			docker buildx build --push \
-				--tag $(IMAGE_NAME):$$arch-$$os-$$tag \
+				--tag $(IMAGE_NAME):$$arch-$$base_image-$$os-$$tag \
```
Comment (pohly) [inline build.make:148]: `base_image` can be empty, leading to a tag with double `--`. Not sure whether it matters, but it would be nicer to avoid this.

**Review by pohly — CHANGES_REQUESTED** [review]
I think you can squash all of this into one commit.

File: build.make (hunk 2 above)
Comment (mauriciopoppe) [reply build.make:148]: correct, I'll mark this as a draft PR because I think it's not ready yet, thanks for checking

Comment (mauriciopoppe) [reply build.make]: fixed

File: build.make (hunk 1 above)
Comment (mauriciopoppe) [reply build.make]: fixed

File: build.make [hunk 3]
```diff
@@ -63,26 +66,35 @@ endif
...
-LDFLAGS = 
+LDFLAGS =
 FULL_LDFLAGS = $(LDFLAGS) $(IMPORTPATH_LDFLAGS) $(EXT_LDFLAGS)
 # This builds each command (= the sub-directories of ./cmd) for the target platform(s)
 # defined by BUILD_PLATFORMS.
 $(CMDS:%=build-%): build-%: check-go-version-go
 	mkdir -p bin
-	echo '$(BUILD_PLATFORMS)' | tr ';' '\n' | while read -r os arch suffix; do \
+	# os_arch_seen captures all of the $$os-$$arch seen for the current binary
+	# that we want to build, if we've seen an $$os-$$arch before it means that
```
Comment (jingxu97) [inline build.make:85]: could you give me an example about this logic?

Comment (mauriciopoppe) [reply build.make:85]: Sure, so if you're iterating over this array of tuples: `linux amd64; linux ppc64le -ppc64le; linux s390x -s390x; linux arm64 -arm64; windows amd64 .exe nanoserver:1809 servercore:ltsc2019; windows amd64 .exe nanoserver:1909 servercore:1909; windows amd64 .exe nanoserver:2004 servercore:2004; windows amd64 .exe nanoserver:20H2 servercore:20H2`

go build will be called like this:

```
+ CGO_ENABLED=0
+ GOOS=linux
+ GOARCH=amd64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=linux
+ GOARCH=ppc64le
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar-ppc64le ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=linux
+ GOARCH=s390x
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar-s390x ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=linux
+ GOARCH=arm64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar-arm64 ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=windows
+ GOARCH=amd64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar.exe ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=windows
+ GOARCH=amd64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar.exe ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=windows
+ GOARCH=amd64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar.exe ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=windows
+ GOARCH=amd64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar.exe ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=windows
+ GOARCH=amd64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar.exe ./cmd/csi-node-driver-registrar
```

As you can see the windows binary is built 4 times for the same os,arch combination, with this code I'm making sure that we only build it once per os,arch combination

```
+ CGO_ENABLED=0
+ GOOS=linux
+ GOARCH=amd64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=linux
+ GOARCH=ppc64le
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar-ppc64le ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=linux
+ GOARCH=s390x
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar-s390x ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=linux
+ GOARCH=arm64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar-arm64 ./cmd/csi-node-driver-registrar
+ CGO_ENABLED=0
+ GOOS=windows
+ GOARCH=amd64
+ go build -a -ldflags ' -X main.version=v12345 -extldflags "-static"' -o ./bin/csi-node-driver-registrar.exe ./cmd/csi-node-driver-registrar
```

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by mauriciopoppe** [comment]
@pohly I made a small change to the name of the image, from `amd64-nanoserver-20H2-windows-canary` to `amd64-windows-nanoserver-20H2-canary, the lgtm tag was removed because of this

**Review by pohly — COMMENTED** [review]
/lgtm

**Comment by mauriciopoppe** [comment]
@pohly so sorry I lost your comment, are the images in csi-driver-host-path still failing because of this?

**Comment by mauriciopoppe** [comment]
@pohly I just checked that you added the missing $ in https://github.com/kubernetes-csi/csi-release-tools/commit/5561780172df4a863139f8677905f112758cea82, thank you and sorry about missing this.

**Pull request merged**